### PR TITLE
chore(security): Make get_database_perm/get_dataset_perm return optional

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

Per [here](https://github.com/apache/superset/blob/97482c57b55c7bc79d923968cdfc06accd9b6788/superset/security/manager.py#L1461-L1468) perms can be `None` and thus these return types (which could be overloaded) should be `Optional[str]` rather than `str`.

### TESTING INSTRUCTIONS

CI.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/security/manager.py [hunk 1]
```diff
@@ -320,11 +320,15 @@ def get_schema_perm(  # pylint: disable=no-self-use
         return None
 
     @staticmethod
-    def get_database_perm(database_id: int, database_name: str) -> str:
+    def get_database_perm(database_id: int, database_name: str) -> Optional[str]:
```
Comment (michael-s-molina) [inline superset/security/manager.py:323]:
```suggestion
    def get_database_perm(database_id: int, database_name: str) -> str | None:
```

File: superset/security/manager.py [hunk 2]
```diff
@@ -320,11 +320,15 @@ def get_schema_perm(  # pylint: disable=no-self-use
...
         return f"[{database_name}].(id:{database_id})"
 
     @staticmethod
-    def get_dataset_perm(dataset_id: int, dataset_name: str, database_name: str) -> str:
+    def get_dataset_perm(
+        dataset_id: int,
+        dataset_name: str,
+        database_name: str,
+    ) -> Optional[str]:
```
Comment (michael-s-molina) [inline superset/security/manager.py:331]:
```suggestion
    ) -> str | None:
```

File: superset/security/manager.py [hunk 3]
```diff
@@ -1271,7 +1283,7 @@ def dataset_after_insert(
         )
 
         try:
-            dataset_perm = target.get_perm()
+            dataset_perm: Optional[str] = target.get_perm()
```
Comment (michael-s-molina) [inline superset/security/manager.py:1286]:
```suggestion
            dataset_perm: str | None = target.get_perm()
```

**Comment by john-bodley** [comment]
@michael-s-molina apologies for the delay in responding to your comments. Regarding replacing `Optional[str]` with `str | None`, currently the `superset/security/manager.py` file doesn't include the, 

```python
from __future__ import annotations
```

import and thus this syntax is not supported. My sense/preference is that this file (alongside others) should be updated in a different PR to keep this change to a minimum.

**Review by michael-s-molina — APPROVED** [review]
LGTM

**Pull request merged**